src/log-store/src/kafka/consumer.rs

```diff
@@ -0,0 +1,380 @@
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        loop {
+            if *this.terminated {
+                return Poll::Ready(None);
+            }
+
+            if this.buffer.index.peek().is_none() {
+                return Poll::Ready(None);
+            }
+
+            if let Some(x) = this.buffer.pop_front() {
+                debug!("Yielding record with offset: {}", x.offset);
+                return Poll::Ready(Some(Ok((x, *this.last_high_watermark))));
+            }
+
+            if this.fetch_fut.is_terminated() {
+                match this.buffer.index.peek() {
+                    Some(next_offset) => {
+                        let client = Arc::clone(this.client);
+                        let max_wait_ms = *this.max_wait_ms as i32;
+                        let offset = next_offset as i64;
+                        let NextBatchHint { bytes, len } = this
+                            .buffer
+                            .index
+                            .next_batch_hint(*this.avg_record_size)
+                            .unwrap_or(NextBatchHint {
+                                bytes: *this.avg_record_size,
+                                len: 1,
+                            });
+
+                        let fetch_range =
+                            1i32..(bytes.saturating_add(1).min(*this.max_batch_size) as i32);
+                        *this.fetch_fut = FutureExt::fuse(Box::pin(async move {
+                            let (records_and_offsets, watermark) = client
+                                .fetch_records(offset, fetch_range, max_wait_ms)
+                                .await?;
+
+                            Ok(FetchResult {
+                                records_and_offsets,
+                                watermark,
+                                used_offset: offset,
+                                fetch_bytes: bytes as i32,
+                                batch_size: len,
+                            })
+                        }));
+                    }
+                    None => {
+                        return Poll::Ready(None);
+                    }
+                }
+            }
+
+            let data = futures::ready!(this.fetch_fut.poll_unpin(cx));
+
+            match data {
+                Ok(FetchResult {
+                    mut records_and_offsets,
+                    watermark,
+                    used_offset,
+                    fetch_bytes,
+                    batch_size,
+                }) => {
+                    // Sort records by offset in case they aren't in order
+                    records_and_offsets.sort_unstable_by_key(|x| x.offset);
+                    *this.last_high_watermark = watermark;
+                    if !records_and_offsets.is_empty() {
+                        *this.avg_record_size = fetch_bytes as usize / records_and_offsets.len();
```

Comment:
I am concerned about significant fluctuations in `avg_record_size`. Here are a few suggestions for improvement:

- Exponential Moving Average (EMA):
```
let alpha = 0.2; // adjustable smoothing factor
*this.avg_record_size = ((1.0 - alpha) * *this.avg_record_size as f64 + 
                        alpha * (fetch_bytes as f64 / records_and_offsets.len() as f64)) as usize;
```

- Incremental Update with Limits:
Set a maximum change magnitude to avoid large single-batch fluctuations.

```
let new_avg = fetch_bytes as usize / records_and_offsets.len();
let max_change = *this.avg_record_size / 10; // allow max 10% change
*this.avg_record_size = (*this.avg_record_size + new_avg.clamp(
    this.avg_record_size.saturating_sub(max_change),
    this.avg_record_size.saturating_add(max_change)
)) / 2;
```

- Sliding Window Average:
Keep a record of the sizes of the last N batches and calculate their average.

```
this.record_sizes.push(fetch_bytes as usize / records_and_offsets.len());
if this.record_sizes.len() > 10 { // keep last 10 records
    this.record_sizes.remove(0);
}
*this.avg_record_size = this.record_sizes.iter().sum::<usize>() / this.record_sizes.len();
```

Reply:
Good point. I'm going to leave it as future work.